### ISSUE TITLE
[Help Wanted] Accessing current values of parameters during an MCMC run

### ISSUE THREAD

**Issue by ms609** [revbayes/revbayes#524]
**Context**
I'm trying to understand the behaviour of a misbehaving model.

Here's some relevant code illustrating one aspect of the model:
```rb
u ~ dnGamma( shape = 1, rate = 1 )
moves.append( mvScale( u ) )
u # Initial value is stochastic

condRates := [ [ 0, u],
               [ 4, 0] ]
Q4cond := fnFreeK(condRates)


m_morph[2] ~ dnInverse( dnPhyloCTMC(tree = phylogeny, Q = Q4cond, type = "Standard")

condChars <- readDiscreteCharacterData("data.nex")
m_morph[2].clamp(condChars)
m_morph[2].lnProbability()
```

I'm keen to sanity-check the relative likelihood values of each component of the model.  Hence I'm doing:

```rb
mymc3 = mcmc(mymodel, monitors, moves, nruns = 2, moveschedule = "single")
for (i in 1:40) {
  mymc3.run(generations = 10) # A small increment
  print( "u = ", u )
  print( "P1 =", m_morph[1].lnProbability(), "P2 =", m_morph[2].lnProbability() )
}
```

My MCMC monitor shows that new values of `u` are being proposed, but the `print()` command only shows the original value of `u`, and the original value of `m_morph[2].lnProbability()`; these are not updating to reflect the current state of the chain.

**Desired outcome**
I'd like to access the current value of `u`, and the current status of `m_morph[2]`, from within RevBayes, either via the console, or by writing calculated values to a monitor 
(e.g. by `mnScreen( u, m_morph[2].lnProbability() )`)

**Comment by davidcerny** [comment]
So for `u`, you should just be able to add it to your monitors, no? Like this:

```
monitors.append( mnModel(filename="mymodel.log", u, printgen=1) )  # print every single iteration
monitors.append( mnScreen(printgen=1, u) )                                             # likewise
```

Since `u` is a stochastic node, it should even be included in the `mnModel` monitor by default, I believe. Perhaps I'm misunderstanding your problem?

When it comes to checking the individual components of your likelihood, your best shot is probably to define a deterministic variable and explicitly tell RevBayes to monitor it:

```
m2_logL := m_morph[2].lnProbability()
monitors.append( mnScreen(printgen=1, u, m2_logL) )
```

Unfortunately, this [doesn't always work](https://github.com/revbayes/revbayes/issues/425).

**Comment by ms609** [comment]
Yes, `u` works in the monitors, but it puzzles me that the value of `u` itself is not updated to match its state in the mcmc object.  I suppose one motivation for not updating its value is that this simplifies the establishment of a second mcmc run from the same starting parameters?

The deterministic variable approach sounds like it can meet my needs, subject to a fix for #425. I assume that issue accounts for why `mymcmc = mcmc(mymodel, monitors, moves)` fails with `cannot find node with name 'm2_logL' in the model but received a monitor working on it`?

**Comment by davidcerny** [comment]
I'm not sure why the value of `u` doesn't get updated (it sounds like @bredelings's conjecture that the chain copies its value instead of modifying it directly is probably correct), but that doesn't necessarily bother me all that much – `print()` is not really meant to monitor the state of a running chain; that's precisely what monitors are for.

The issue with the deterministic node not being found unless it is explicitly included in the `model()` function sounds more serious, and it might be worth reporting it as a separate bug.

**Comment by ms609** [comment]
OK – the "model works with a copy" explanation makes sense; I've proposed noting this in the documentation at #534.

I've reported a further instance of parameters not being included in the model in #533.

Once #425 is fixed I think that solves my initial problem, so I'll close this issue now.
